### ISSUE TITLE
using a GridSampler with a Queue

### ISSUE THREAD

**Issue by bcdarwin** [fepegar/torchio#127]
Right now this doesn't exactly make sense, e.g. you need to pass `samples_per_volume` to `Queue` which is not quite coherent, and it's not clear how to pass arguments (specifically `patch_overlap`) to the `GridSampler`.

(Use case: I want my validation to be done deterministically and from the entire image which is however too big to be segmented in a single pass, even without gradients.  I guess there are also some minor issues with computing validation metrics on the patches rather than after stitching the patches but this is a start.)

**Comment by fepegar** [comment]
[Have you tried the `GridAggregator`?](https://torchio.readthedocs.io/data/sampling.html#sampling)

**Comment by bcdarwin** [comment]
Yes, but I want to use the `GridSampler` with a `Dataset`, not a fixed input tensor so (I think) I want `Queue`-like functionality.

**Comment by fepegar** [comment]
Unfortunately, this is not well documented yet (#97).

I don't think you want a queue for this. I think the best would be iterating over the dataset samples and create a `GridAggregator` for each tensor. A bit like what's done in [this section of the notebook](https://colab.research.google.com/drive/112NTL8uJXzcMw4PQbUvMQN-WHlVwQS3i#scrollTo=SKu8k5Z2wkSD&line=6&uniqifier=1). I guess you can put in a function most of the code in that cell and call the function for each sample. Would that work for you?

I'll add more examples soon.

**Comment by bcdarwin** [comment]
To be clear, what I am trying to do is compute validation statistics at each training epoch.  I don't really need to assemble the whole volume for this, but I do want it to be fast.  Is the above still what you suggest?

**Comment by fepegar** [comment]
If you don't need to assemble the volume, you can still use the code from the notebook minus the aggregator bits. Here's some pseudocode:

```
metrics = {}
for each sample in validation dataset:
    get tensor from sample
    sampler = grid_sampler(tensor)
    metrics_subject = []
    for patch in sampler:
        patch_x, patch_y = prepare_patch(patch)
        metric = get_metric(patch_x, patch_y)
        metrics_subject.append(metric)
    metrics[subject] = metrics_subject
```

---

After thinking a bit more and looking at the code, I think the problem with the above is that the GridSampler should take a sample as input, not just a tensor, as for the other samplers. I'll see what I can do and come up with a solution.

**Comment by bcdarwin** [comment]
> I think the problem with the snippet above is that the GridSampler should take a sample as input, not just a tensor, like the other samplers.

Exactly, it seems like for the GridAggregator to work the GridSampler should produce a nested dataset of lists (or arrays or datasets) of patches?

**Comment by fepegar** [comment]
Would this work for you?
https://github.com/fepegar/torchio/blob/0903f6ae91818a652555a3aa2670333054bcc2db/tests/data/test_inference.py#L19-L27

**Comment by fepegar** [comment]
Out of curiosity, why do you compute the metrics patch by patch? What is the task, if I may ask?

**Comment by bcdarwin** [comment]
It's for segmentation of large MRI volumes.  Training has to be done patchwise as per your Colab notebook or the GPU runs out of RAM.  I just wanted to do inference in a more deterministic way so I thought I could plug in GridSampler instead of ImageSampler in the validation phase of each training epoch.  I'm not committed to this approach but I think even running the network forward has to be done in patches for RAM reasons.

**Comment by fepegar** [comment]
Can you take a look at the snippet in my comment? https://github.com/fepegar/torchio/issues/127#issuecomment-618052711

**Comment by bcdarwin** [comment]
In principle that (manually iterating over the data) would work, but in that case why not also run the training loop that way (i.e., without using a Queue)?

**Comment by fepegar** [comment]
That can certainly be done. But the queue allows you to automatically load the data in the background, randomize patches within the queue and make sure that they come from different subjects for each patch, etc.

**Comment by bcdarwin** [comment]
I guess I don't need want/need the randomization but I would prefer to have multithreaded data loading/preprocessing?

**Comment by fepegar** [comment]
You could do something like this:

```python
import torchio
from torch.utils.data import DataLoader

patch_size = 64
patch_border = 4
training_batch_size = 6

training_set = torchio.datasets.IXITiny('/tmp/ixi')

training_loader = DataLoader(
    training_set,
    batch_size=1,
    num_workers=4,
    collate_fn=lambda x: x[0],
)

for sample in training_loader:
    sampler = torchio.data.GridSampler(sample, patch_size, patch_border)
    patch_loader = DataLoader(sampler, batch_size=training_batch_size)
    for batch in patch_loader:
        inputs = batch['image'][torchio.DATA]
        targets = batch['label'][torchio.DATA]
        run_train_iteration(inputs, targets)

```

**Comment by fepegar** [comment]
@all-contributors please add @bcdarwin for ideas
